crates/polars-stream/src/physical_plan/lower_ir.rs

```diff
@@ -314,23 +314,93 @@ pub fn lower_ir(
                 sources: scan_sources,
                 file_info,
                 hive_parts,
-                output_schema,
+                output_schema: scan_output_schema,
                 scan_type,
-                predicate,
-                file_options,
+                mut predicate,
+                mut file_options,
             } = v.clone()
             else {
                 unreachable!();
             };
 
-            PhysNodeKind::FileScan {
+            // # Operation ordering
+            // with_row_index() -> slice() -> filter()
+
+            // Some scans have built-in support for applying these operations in an optimized manner.
+            let opt_rewrite_to_nodes = match &scan_type {
+                FileScan::Parquet { .. } => (None, None, None),
+                FileScan::Csv { options, .. } => {
+                    if file_options.slice.map_or(true, |(offset, _)| {
+                        // Only really makes sense to push this if we are skpping rows
+                        offset > 0
+                    }) && options.parse_options.comment_prefix.is_none()
+                        && std::env::var("POLARS_NEW_STREAMING_CSV_EXPERIMENTAL_SLICE").as_deref()
+                            == Ok("1")
```

Comment:
I have slice pushdown into the CSV source disabled by default - it relies heavily on `CountLines` being correct.